x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Analyzer.java

```diff
@@ -42,4 +50,56 @@ private LogicalPlan verify(LogicalPlan plan) {
         }
         return plan;
     }
-}
+
+    private static class ResolveRefs extends AnalyzeRule<LogicalPlan> {
+
+        @Override
+        protected LogicalPlan rule(LogicalPlan plan) {
+            // if the children are not resolved, there's no way the node can be resolved
+            if (!plan.childrenResolved()) {
+                return plan;
+            }
+
+            // okay, there's a chance so let's get started
+            if (log.isTraceEnabled()) {
+                log.trace("Attempting to resolve {}", plan.nodeString());
+            }
+
+            return plan.transformExpressionsUp(e -> {
+                if (e instanceof UnresolvedAttribute) {
+                    UnresolvedAttribute u = (UnresolvedAttribute) e;
+                    List<Attribute> childrenOutput = new ArrayList<>();
+                    for (LogicalPlan child : plan.children()) {
+                        childrenOutput.addAll(child.output());
+                    }
+                    NamedExpression named = resolveAgainstList(u, childrenOutput);
+                    // if resolved, return it; otherwise keep it in place to be resolved later
+                    if (named != null) {
+                        if (log.isTraceEnabled()) {
+                            log.trace("Resolved {} to {}", u, named);
+                        }
+                        return named;
+                    }
+                }
+                return e;
+            });
+        }
+    }
+
+    abstract static class AnalyzeRule<SubPlan extends LogicalPlan> extends Rule<SubPlan, LogicalPlan> {
```

Comment:
There will be a separate PR to externalize this to QL so it can be shared across SQL and EQL.